Wikipedia/Code/NotificationsCenterViewModel+LinkExtensions.swift

```diff
@@ -0,0 +1,263 @@
+
+    }
+}
+
+//MARK: Private Helpers - URL Generation Methods
+
+private extension NotificationsCenterCellViewModel {
+
+    /// Generates a wiki url with the full (i.e. already prefixed) title from the notification
+    func fullTitleURL(for configuration: Configuration) -> URL? {
+        guard let data = linkData(for: configuration),
+              let fullTitle = data.fullTitle else {
+            return nil
+        }
+
+        guard let url = configuration.articleURLForHost(data.host, languageVariantCode: data.languageVariantCode, appending: [fullTitle]) else {
+            return nil
+        }
+
+        guard let namespace = data.titleNamespace,
+              namespace == .userTalk,
+              var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            return url
+        }
+
+        components.fragment = data.primaryLinkFragment
+        return components.url
+    }
+    
+    /// Generates a wiki diff url with the full (i.e. already prefixed) title from the notification
+    func fullTitleDiffURL(for configuration: Configuration) -> URL? {
+        guard let data = linkData(for: configuration),
+              let fullTitle = data.fullTitle,
+              let revisionID = data.revisionID else {
+            return nil
+        }
+
+        guard let url = configuration.expandedArticleURLForHost(data.host, languageVariantCode: data.languageVariantCode, queryParameters: ["title": fullTitle, "oldid": revisionID]) else {
+            return nil
+        }
+
+        return url
+    }
+    
+    var primaryLinkMinusQueryItemsURL: URL? {
+        guard let url = notification.primaryLinkURL,
+              var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            return nil
+        }
+
+        components.queryItems?.removeAll()
+        return components.url
+    }
+    
+    //https://www.mediawiki.org/wiki/Special:UserGroupRights
+    var userGroupRightsURL: URL? {
+        //Note: Sample notification json indicates that translated user group link we want is listed as the primary URL
+        //Ex. https://en.wikipedia.org/wiki/Special:ListGroupRights?markasread=nnnnnnnn&markasreadwiki=enwiki#confirmed
+        
+        guard let url = primaryLinkMinusQueryItemsURL,
+              var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            return nil
+        }
+
+        components.fragment = nil
+        return components.url
+    }
+    
+    var connectionWithWikidataItemURL: URL? {
+
+        //Note: Sample notification json indicates that the wikidata item link is the second secondary link.
+        //Return this link if we're fairly certain it's what we think it is
+        
+        guard let secondaryLinks = notification.secondaryLinks,
+              secondaryLinks.count > 1,
```

Comment:
This is fine as is, but just wanted to mention that generally for safe indexing purposes I'd favor a technique that directly checks the availability of that index (`secondaryLinks.indicies.contains(1)`) rather than relying on a secondary property (`count`).

Reply:
@staykids good technique, the intention feels more clear with that way. This is updated.